File: coreNodes/Node.py
#!/usr/bin/env python         
# -*- coding: utf-8 -*-

import time
import serial
from Agreement import Agreement

class Node(object):
    '''
    id_;//设备编号，固定8位二进制
    function_;//提供的功能，固定16位二进制
    length_;  //表长度，保存网络中当前节点的个数
    routingList_;//路由表，保存网络中的设备地址 8位二进制
    functionList_;//功能表，保存网络中，某个设备的功能编号 16位二进制

    Agreement agreement = Agreement();//协议对象，用于通讯的解码和编码
    '''

    def __init__(self, id, func, portname, baudrate = 9600):
        #初始化，需要节点编号，功能编号，串口名称，与波特率信息
        self.__id = id
        self.__function = func
        self.__length = 0
        self.__routingList = []
        self.__functionList = []
        self.__agreement = Agreement()
        try:
            self.__serialPort = serial.Serial(portname, baudrate)
            self.__serialPort.writeTimeout = 2 #写超时
            print ("串口打开成功%s"%portname),
            print (",波特率为%s"%baudrate)
        except Exception as e:
            print ("串口被占用，请重新启动")


    def getSerialPort(self):
        #返回串口对象
        print ("getSerialPort")
        return self.__serialPort
    

    def update(self):
        #主节点将会发送刷新命令，将会把所有的路由表、功能表都清空，长度设置为0
        print ("update")
        self.__length = 0
        self.__routingList = []
        self.__functionList = []
        #给type字段设置为 0x00编号
        self.__agreement.source = self.__id
        self.__agreement.destination = 0xff
        self.__agreement.type = 0x00
        self.__agreement.data_0 = 0x00
        self.__agreement.data_1 = 0x00
        #发送消息
        self.__serialPort.write(self.__agreement.encode())
        time.sleep(0.01)


    def addList(self):
        #主节点收到设备的信息(0x01)后，先把信息添加到主节点的表中，然后将会把设备的信息添加到表中，并广播给所有的节点(0x02)
        print ("addList")
        if self.__agreement.source in self.__routingList:
            #如果表中已经有这个设备的信息的话，就不再添加
            pass
        else:
            #添加到表中
            self.__length += 1
            self.__routingList.append(self.__agreement.source)
            self.__functionList.append((self.__agreement.data_0 << 4) + self.__agreement.data_1 )
            #广播信息，更新到别的设备
            self.__agreement.type = 0x02
            self.translate()
    

    def sendInfo(self):
        #将主节点的设备信息发布出去
        print ("sendInfo")
        self.__agreement.source = self.__id
        self.__agreement.destination = 0xff
        self.__agreement.type = 0x02
        self.__agreement.data_0 = self.__function & 0x1100
        self.__agreement.data_1 = self.__function & 0x0011
        self.__serialPort.write(self.__agreement.encode())
        time.sleep(0.01)
    

    def translate(self):
        #数据转发，当收到sendData之后，将数据转发走
        print ("translate")
        self.__serialPort.write(self.__agreement.encode())
        time.sleep(0.01)


    def listening(self, s):
        #监听，并判断该执行什么任务
        self.__agreement.decode(s)
        if( self.__agreement.type == 0x01 ):
            #添加路由表
            self.addList()
        elif( self.__agreement.type == 0xff ):
            #转发数据
            self.translate()
        else:
            print("其他消息")
            
    
    def printInfo(self):
        #打印出当前节点状态
        print("\n\nid %s,"%self.__id),
        print("function %s"%self.__function)
        print("routingList :"),
        for i in range(0, self.__length):
            print(self.__routingList[i]),
        print("\nfunctionList :"),
        for i in range(0, self.__length):
            print(self.__routingList[i]),
        print("")

File: libraries/Node/Node.cpp
#include "Node.h"

//节点初始化
Node::Node(int id, int func) : id_(id), function_(func)
{
    //std::cout << "Node\n";
    length_ = 0;
    update();
    Serial.begin(9600);
}

/*
void Node::printInfo() const
{
    std::cout << id_ << " " << length_ << " " << function_ << std::endl;
    for(int i = 0 ; i < length_ ; i ++)
    {
        std::cout << routingList_[i] << " " << functionList_[i] << std::endl;
    }
}
*/

//0x00 收到刷新命令，因为代价比较小，所以定时刷新，并重新建表
void Node::update()
{
    //std::cout << "update\n";
    if(length_ > 0)
    {
        delete routingList_;
        delete functionList_;
        length_ = 0;
    }
    routingList_ = NULL;
    functionList_ = NULL;
}


//0x01 收到刷新组网指令后，执行 发送信息指令
void Node::sendInfo()
{
    //std::cout << "sendInfo\n";
    agreement.source = id_;
    agreement.destination = 0x00;
    agreement.type = 0x01;
    agreement.data_0 = function_ >> 8;
    agreement.data_1 = function_ & 0x00ff;
    //std::cout << agreement.encode() << std::endl;
    const char *s = agreement.encode();
    for(int i = 0 ; i < agreement.sumSize ; i ++)
    {
        Serial.write(s[i]);
    }
    Serial.println();
    delete s;
    s = NULL;
}


//0x02 收到添加路由表指令后，执行
void Node::addList()
{
    //std::cout << "addList\n";

    bool flag = true;
    //表中不应包含自身节点的节点号
    if(agreement.source == id_ )
    {
        flag = false;
    }
    //检查是否重复
    for(int i = 0 ; i < length_; i++)
    {
        if(agreement.source == routingList_[i])
        {
            flag = false;
        }
    }

    if(flag)
    {
        //申请新的指针以及内存空间
        int *rp = new int[length_ + 1];
        int *fp = new int[length_ + 1];

        for(int i = 0; i < length_; i++)
        {
            //复制之前的数据
            rp[i] = routingList_[i];
            fp[i] = functionList_[i];
        }
        //在表末尾添加最新数据
        rp[length_] = agreement.source;
        fp[length_] = (agreement.data_0 << 8 )+ agreement.data_1;

        //改变表长
        length_ += 1;

        //删除之前的表
        delete routingList_;
        delete functionList_;

        //把指针指向新的内存地址
        routingList_ = rp;
        functionList_ = fp;
    }

    /*
    std::cout << "rout func\n";
    for(int i = 0 ; i < length_ ; i ++)
    {
        std::cout << routingList_[i] << " " << functionList_[i] << std::endl;
    }
    */
}


//监听消息，并解析消息来委派的组网任务，如果有特殊任务，那么返回true，正常任务返回false
bool Node::listening(unsigned char *s)
{
    //用协议解码
    agreement.decode(s);
    //根据type字段来进行判断到底执行哪些动作
    switch(agreement.type)
    {
    case 0x00:
        //刷新
        update();
        //发送节点信息
        sendInfo();
        break;
    case 0x02:
        //向表里添加信息
        addList();
        break;
    default:
        //属于正常任务
        return false;
    }
    return true;
}


void Node::sendData(int data)
{
    //std::cout << "sendData\n";
    agreement.source = id_;
    agreement.type = 0xff;
    agreement.data_0 = data >> 8;
    agreement.data_1 = data & 0x00ff;
    for(int i = 0 ; i < length_; i++)
    {
        if( function_ & functionList_[i] > 0 )
        {
            agreement.destination = routingList_[i];
            //std::cout << agreement.encode() << std::endl;
            const char *s = agreement.encode();
            for(int i = 0 ; i < agreement.sumSize ; i ++)
            {
                Serial.write(s[i]);
            }
            Serial.println();
            delete s;
            s = NULL;
        }
    }
}


void Node::sendData(int data, int destination)
{
    //std::cout << "sendData\n";
    agreement.source = id_;
    agreement.destination = destination;
    agreement.type = 0xff;
    agreement.data_0 = data >> 8;
    agreement.data_1 = data & 0x00ff;
    const char *s = agreement.encode();
    for(int i = 0 ; i < agreement.sumSize ; i ++)
    {
        Serial.write(s[i]);
    }
    Serial.println();
    delete s;
    s = NULL;
}


//过滤器，如果消息源头有效，那么返回true。否则返回false
bool Node::filter() const
{
    if(getAgreement().source != id_)
    {//判断是否出现链路回传。如果是回传，那么拒收消息
        if((getAgreement().destination == id_ )|| (getAgreement().destination == 0xff))
        {//如果消息的目标是本节点编号，或者是多播类型，那么接收消息
            return true;
        }
    }
    return false;
}

//测试


Node::~Node()
{
}
